src/server/scripts/Outland/HellfireCitadel/ShatteredHalls/instance_shattered_halls.cpp

```diff
@@ -100,24 +187,309 @@ class instance_shattered_halls : public InstanceMapScript
+
+                Initialize();
+            }
+
+            void JustSummoned(Creature*) override { }
+
+            void JustDied(Unit*) override
+            {
+                _JustDied();
+                Map::PlayerList const &players = instance->instance->GetPlayers();
+                for (Map::PlayerList::const_iterator itr = players.begin(); itr != players.end(); ++itr)
+                {
+                    Player* pl = itr->GetSource();
+                    uint32 qId = pl->GetTeam() == ALLIANCE ? QUEST_IMPRISONED_A : QUEST_IMPRISONED_H;
+                    if (pl->GetQuestStatus(qId) != QUEST_STATUS_NONE && pl->GetQuestStatus(qId) != QUEST_STATUS_FAILED)
+                        pl->CompleteQuest(qId);
+                }
+            }
+
+            void SetData(uint32 type, uint32 data) override
+            {
+                if (type == 1)
+                {
+                    uint32 entry = executionerVictims[data - 1](instance->GetData(DATA_TEAM_IN_INSTANCE));
+                    if (Creature* victim = me->FindNearestCreature(entry, 30.0f, true))
+                        me->Kill(victim);
+
+                    if (data == 1)
+                    {
+                        Map::PlayerList const &players = instance->instance->GetPlayers();
+                        for (Map::PlayerList::const_iterator itr = players.begin(); itr != players.end(); ++itr)
+                        {
+                            Player* pl = itr->GetSource();
+                            uint32 qId = pl->GetTeam() == ALLIANCE ? QUEST_IMPRISONED_A : QUEST_IMPRISONED_H;
+                            if (pl->GetQuestStatus(qId) == QUEST_STATUS_INCOMPLETE)
+                                pl->FailQuest(qId);
+                        }
+                    }
+
+                    switch (instance->GetData(DATA_PRISONERS_EXECUTED))
+                    {
+                        case 3:
+                            me->RemoveLootMode(LOOT_MODE_HARD_MODE_1);
+                        case 2:
+                            me->RemoveLootMode(LOOT_MODE_HARD_MODE_2);
+                        case 1:
+                            me->RemoveLootMode(LOOT_MODE_HARD_MODE_3);
+                        default:
+                            break;
+                    }
+                }
+            }
+
+            void UpdateAI(uint32 diff) override
+            {
+                if (!UpdateVictim())
+                    return;
+
+                if (cleaveTimer <= diff)
+                {
+                    DoCast(SPELL_CLEAVE);
+                    cleaveTimer = urand(5000, 7000);
+                }
+                else
+                    cleaveTimer -= diff;
+
+                DoMeleeAttackIfReady();
+            }
+        private:
+            InstanceScript* instance;
+            uint32 cleaveTimer;
+        };
+
+        CreatureAI* GetAI(Creature* creature) const override
+        {
+            return new boss_shattered_executionerAI(creature);
```

Comment:
should this be
```
return GetInstanceAI<boss_shattered_executionerAI>(creature);
```
?
to make sure creature is in a map with an InstanceScript. I'm not sure if there's a special function for BossAI too